### ISSUE TITLE
Linter: Error when setting initial state in constructor of class component

### ISSUE THREAD

**Issue by marvinhagemeister** [oxc-project/oxc#3290]
The `eslint-plugin-react(no-direct-mutation-state)` linting rule doesn't seem to differentiate between methods and the `constructor` method. Mutating `this.state` in `constructor` is fine and a common pattern to set the initial state.

```ts
import { Component } from "preact";

class Foo extends Component{
  constructor() {
    super();
    this.state = { i: 1 };
  }
  componentDidMount() {
    doRender = () => this.setState({ i: 2 });
  }
  render(props, { i }) {
    return <WithDefaultProps fieldA={1} fieldB={i} fieldD={i} />;
  }
}
```

This pattern should be correct as indicated in the original linting rule here https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-direct-mutation-state.md

**Comment by JoviDeCroock** [comment]
The issue doesn't reproduce for me with the example, it does however show on https://github.com/preactjs/preact/blob/main/test/browser/spec.test.js#L41-L52

**Comment by marvinhagemeister** [comment]
Whoops, yeah updated the example snippet.

**Comment by JoviDeCroock** [comment]
It's pretty odd, even that snippet does not reproduce in tests when I run them. Trying to investigate how that's possible

**Comment by Boshen** [comment]
Fixed in v0.3.5 https://github.com/oxc-project/oxc/releases/tag/oxlint_v0.3.5
